[Test Update] Updating TC_CADMIN_1_3_4 and cadmin_support python3 modules for clock skew allowance subscription

j-ororke opened the pull request:
#### Summary
- The change here adds subscription monitoring of the DUT's commissioning windows status for a set amount of time with addition clock skew allowance of 100ms or 1% of time allotted for commissioning window to be open whichever is greater, this removes needing to have hard coded sleeps in the test code and allows us to know more accurately of when the window status changes on the DUT.

#### Related issues
- This fixes issue #[610](https://github.com/project-chip/matter-test-scripts/issues/610)
- Test change PR here: [5276](https://github.com/CHIP-Specifications/chip-test-plans/pull/5276)

#### Testing
- TC_CADMIN_1_3_4 python3 test module has implemented the subscription window monitoring method from cadmin_support module in test step 9 for the 180s timed commissioning session for functional validation

#### Results from testing
```
WSL Result:
[2025-06-27 10:55:37.477927][TEST][STDOUT][MatterTest] 06-27 10:55:37.477 INFO === COMMISSIONING WINDOW TIMING RESULTS ===
[2025-06-27 10:55:37.477998][TEST][STDOUT][MatterTest] 06-27 10:55:37.477 INFO Window closed: ✅ YES
[2025-06-27 10:55:37.478024][TEST][STDOUT][MatterTest] 06-27 10:55:37.477 INFO Timing valid: ✅ YES
[2025-06-27 10:55:37.478043][TEST][STDOUT][MatterTest] 06-27 10:55:37.477 INFO ⏱️  TIMING BREAKDOWN:
[2025-06-27 10:55:37.478090][TEST][STDOUT][MatterTest] 06-27 10:55:37.477 INFO    Expected duration: 180s
[2025-06-27 10:55:37.478124][TEST][STDOUT][MatterTest] 06-27 10:55:37.477 INFO    Actual duration: 180.09s
[2025-06-27 10:55:37.478158][TEST][STDOUT][MatterTest] 06-27 10:55:37.477 INFO    Clock skew applied: 1800ms
[2025-06-27 10:55:37.478191][TEST][STDOUT][MatterTest] 06-27 10:55:37.477 INFO    Maximum allowed: 181.80s
[2025-06-27 10:55:37.478233][TEST][STDOUT][MatterTest] 06-27 10:55:37.477 INFO    ⚠️  Window closed LATE but within tolerance by 0.09s
[2025-06-27 10:55:37.478291][TEST][STDOUT][MatterTest] 06-27 10:55:37.478 INFO    Start time: 2025-06-27 10:52:37.391457
[2025-06-27 10:55:37.478327][TEST][STDOUT][MatterTest] 06-27 10:55:37.478 INFO    End time: 2025-06-27 10:55:37.477659
[2025-06-27 10:55:37.478365][TEST][STDOUT][MatterTest] 06-27 10:55:37.478 INFO    Total monitoring time: 180.09s
[2025-06-27 10:55:37.478413][TEST][STDOUT][MatterTest] 06-27 10:55:37.478 INFO === END TIMING RESULTS ===

TH Docker Results:
[MatterTest] 06-27 23:06:57.747 INFO ⏱️  TIMING BREAKDOWN:
[MatterTest] 06-27 23:06:57.748 INFO    Expected duration: 180s
[MatterTest] 06-27 23:06:57.748 INFO    Actual duration: 180.03s
[MatterTest] 06-27 23:06:57.749 INFO    Clock skew applied: 1800ms
[MatterTest] 06-27 23:06:57.749 INFO    Maximum allowed: 181.80s
[MatterTest] 06-27 23:06:57.749 INFO    ⚠️  Window closed LATE but within tolerance by 0.03s
[MatterTest] 06-27 23:06:57.750 INFO    Start time: 2025-06-27 23:03:57.714978
[MatterTest] 06-27 23:06:57.750 INFO    End time: 2025-06-27 23:06:57.746295
[MatterTest] 06-27 23:06:57.751 INFO    Total monitoring time: 180.03s
[MatterTest] 06-27 23:06:57.751 INFO === END TIMING RESULTS ===

TH UI Results:
[MatterTest] 06-27 23:29:00.464 INFO === COMMISSIONING WINDOW TIMING RESULTS ===
[MatterTest] 06-27 23:29:00.467 INFO ⏱️  TIMING BREAKDOWN:
[MatterTest] 06-27 23:29:00.470 INFO    Actual duration: 180.08s
[MatterTest] 06-27 23:29:00.471 INFO    Clock skew applied: 1800ms
[MatterTest] 06-27 23:29:00.472 INFO    Maximum allowed: 181.80s
[MatterTest] 06-27 23:29:00.473 INFO    ⚠️  Window closed LATE but within tolerance by 0.08s
[MatterTest] 06-27 23:29:00.474 INFO    Start time: 2025-06-27 23:26:00.382066
[MatterTest] 06-27 23:29:00.475 INFO    End time: 2025-06-27 23:29:00.464464
[MatterTest] 06-27 23:29:00.476 INFO    Total monitoring time: 180.08s
[MatterTest] 06-27 23:29:00.478 INFO === END TIMING RESULTS === 
```

j-ororke committed:
Updating TC_CADMIN_1_3_4 and cadmin_support python3 modules:
- Updated these to include new real-time commissioning window monitoring with subscriptions to remove prior hardcoded sleep
- Added timing validation with clock skew tolerance
- Provides detailed timing metrics to show exactly how long the commissioning window was open and how much clock skew applied
- Removed test steps 10 and 11 as no longer using in TC_CADMIN_1_3_4 python3 test module, as we get the window closed status back in our subscription at the end of test step 9.

j-ororke committed:
isort and autopep8 stylizered

j-ororke committed:
Resolving broad exception, changed to asyncio.TimeoutError instead

j-ororke commented on src/python_testing/support_modules/cadmin_support.py:
```diff
@@ -68,6 +71,268 @@ async def get_txt_record(self):
+        except Exception as e:
+            logging.error(f"❌ Timeout waiting for window status {expected_status} ({status_name}): {e}")
+            logging.error(f"Timeout occurred after {timeout_sec}s")
+            return False
```
Thank you, this has been implemented and fixed now

j-ororke committed:
Resolving linting errors

tcarmelveilleux commented on src/python_testing/support_modules/cadmin_support.py:
```diff
@@ -68,6 +71,267 @@ async def get_txt_record(self):
+        Returns:
+            Clock skew factor in milliseconds (1% of duration or 100ms, whichever is greater)
+        """
+        skew_percentage = 0.01  # 1%
```
```suggestion
        skew_percentage = 1 / 100.0
```

j-ororke replied:
Updated, thank you!

tcarmelveilleux commented on src/python_testing/support_modules/cadmin_support.py:
```diff
@@ -68,6 +71,267 @@ async def get_txt_record(self):
+            if not window_closed:
+                timing_valid = False
+                logging.error("❌ WINDOW DID NOT CLOSE WITHIN EXPECTED TIME")
+                logging.error(f"Monitoring timed out after {max_allowed_duration + 10}s")
```
Why "+ 10" and not "max_allowed_duration" ?

Suggest refactoring the timeout of "+10" into a new helper variable and using it directly, so there is one source of truth.

j-ororke replied:
I have updated the logic here to create a new variable named monitoring_timeout which is max_allowed_duration + 10 seconds.

**Why max_allowed_duration + 10 seconds?**
The +10 seconds is a pragmatic, conservative buffer to ensure the test logic has enough time to observe the status change, accounting for real-world delays in the subscription/reporting pipeline. It does not affect the actual pass/fail criteria for the DUT's timing accuracy.

**Key points:**
- The buffer accounts for network latency, thread scheduling, and subscription processing delays
- The actual timing validation still uses max_allowed_duration (expected duration + clock skew)
- This prevents false test failures due to late status reports while maintaining strict timing requirements for the DUT

j-ororke commented on src/python_testing/support_modules/cadmin_support.py:
```diff
@@ -68,6 +71,267 @@ async def get_txt_record(self):
+    async def wait_for_window_status_change(
+        self,
+        window_status_accumulator: ClusterAttributeChangeAccumulator,
+        expected_status: int,
```
Good point, I have updated the code here now to using a bool value for is_open_expected var in place of an int value for the prior var expected_status.

j-ororke committed:
Updating cadmin_support python3 support module:
- Created a new variable named monitoring_timeout which is max_allowed_duration + 10 seconds.
- Updated the commissioning option var name to commissioning_option for clarity and now use a local CommissioningWindowOption enum as well as updated docstring
- Updated expected_status from an int to a True or False bool value, changed expected_status to is_open_expected var name

j-ororke commented on src/python_testing/support_modules/cadmin_support.py:
```diff
@@ -68,6 +71,267 @@ async def get_txt_record(self):
+            timeout: Window timeout in seconds
+            node_id: Target node ID
+            discriminator: Optional discriminator value
+            option: Commissioning option (default: 1)
```
Thank you for the feedback!
I've updated the parameter name from option to commissioning_option for clarity and now use a local CommissioningWindowOption enum. The docstring has also been expanded to clearly describe the valid values and their meanings, making the code more self-documenting and type-safe.

restyled-commits committed:
Restyled by autopep8

restyled-commits committed:
Restyled by isort

j-ororke committed:
Updating cadmin_support python3 support module:
- Changing "if is_open_expected=False" to "if not is_open_expected" in order to resolve linting error.

j-ororke committed:
Merge branch 'master' into create_CADMIN_clock-skew

j-ororke committed:
Merge branch 'master' into create_CADMIN_clock-skew

j-ororke committed:
Update cadmin_support.py

Changing import location of ClusterAttributeChangeAccumulator from chip.testing.matter_testing to chip.testing.event_attribute_reporting

raul-marquez-csa commented on src/python_testing/support_modules/cadmin_support.py:
```diff
@@ -68,13 +78,278 @@ async def get_txt_record(self):
         )
         return comm_service
```
`get_txt_record` can be removed as it's unused.

j-ororke replied:
Thank you Raul, 
I have removed the unused get_txt_record function from the cadmin_support support module!

j-ororke committed:
remove get_txt_record function from cadmin_support support module as it is unused

j-ororke committed:
Merge branch 'master' into create_CADMIN_clock-skew

restyled-commits committed:
Restyled by autopep8

restyled-commits committed:
Restyled by isort

j-ororke committed:
Merge branch 'master' into create_CADMIN_clock-skew

j-ororke committed:
Merge branch 'master' into create_CADMIN_clock-skew

cecille commented on src/python_testing/support_modules/cadmin_support.py:
```diff
@@ -58,13 +73,340 @@ async def read_currentfabricindex(self, th: ChipDeviceCtrl) -> int:
+
+            # Interpret integers/enums: 0 == closed, 1 == open, 2 = Busy
+            if isinstance(report_value, int):
+                return report_value != 0
```
Can you use the CommissioningWindowStatusEnum here?

j-ororke replied:
Hi Cecille,
My apologies as mentioned during this mornings team meeting when the merge happened with AttributeChangeCallback and ClusterAttributeChangeAccumulator classes into AttributeSubscriptionHandler class which was in PR #39672, which you had me review I had verified the changes would not effect the currently merged CADMIN tests, but I had not realized what it would do to our in-progress clock-skew PR here. 
However, I have now restored the original code which now includes the clock tolerance functionality and refactored the code to now pass with all CADMIN tests, resolving the prior conflicts noticed after merging in the latest changes from master into this PR branch.  
The above change you have mentioned in this comment is no longer implemented as the restored code had already been reviewed by Raul and Tennessee which seemed much more simple, linear, and easier to understand, hopefully making it easier to maintain moving forward than the new code that I had been attempting to implement which you comment references.

cecille commented on src/python_testing/support_modules/cadmin_support.py:
```diff
@@ -58,13 +73,340 @@ async def read_currentfabricindex(self, th: ChipDeviceCtrl) -> int:
+            if isinstance(report_value, int):
+                return report_value != 0
+
+            # Fall back: best-effort equality/semantics
```
when does this happen?

j-ororke committed:
Updating cadmin_support support and TC_CADMIN_1_3_4 test modules:
- Due to merge of AttributeChangeCallback and ClusterAttributeChangeAccumulator classes into AttributeSubscriptionHandler class in PR #39672 had to make some changes to get this working after that PR merged
- Resolved some minor overlap with other CADMIN tests and some of the support functions for opening commissioning window in cadmin_support support module
- Verified all CADMIN test modules are now passing

restyled-commits committed:
Restyled by isort

j-ororke commented on src/python_testing/support_modules/cadmin_support.py:
```diff
@@ -58,13 +73,306 @@ async def read_currentfabricindex(self, th: ChipDeviceCtrl) -> int:
+        try:
+            window_status_accumulator.await_all_expected_report_matches([status_match], timeout_sec=timeout_sec)
+            logging.info(f"✅ Window status changed to {status_name} (status={is_open_expected})")
+            return True
```
Ah good point, thank you for catching this. I have updated this function to now only asserting if the timeout error is experienced, otherwise it silently continues once the window is closed. This keeps the timeout error assert immediately.

j-ororke commented on src/python_testing/support_modules/cadmin_support.py:
```diff
@@ -58,13 +73,306 @@ async def read_currentfabricindex(self, th: ChipDeviceCtrl) -> int:
+
+            # Verify timing
+            timing_valid = True
+            if not window_closed:
```
You're absolutely right! I've removed the dead code from the log_timing_results function as well. 
Since window_closed is now always True (we only reach that point if the window actually closed), I've simplified the logging to remove the redundant checks and always show 'Window closed: ✅ YES'

j-ororke commented on src/python_testing/support_modules/cadmin_support.py:
```diff
@@ -58,13 +73,306 @@ async def read_currentfabricindex(self, th: ChipDeviceCtrl) -> int:
+            # Clean up subscription
+            await window_status_accumulator.cancel()
+
+    async def open_commissioning_window_with_full_args(
```
Ah yes, since this is really only called once and we did start using it again as intended, I have now moved the code contained in this function to inside of the open_commissioning_window_with_subscription_monitoring function to remove the unneeded extra code that this function was creating. Trying to keep it simple, making it easier to maintain moving forward.

j-ororke commented on src/python_testing/support_modules/cadmin_support.py:
```diff
@@ -58,13 +73,306 @@ async def read_currentfabricindex(self, th: ChipDeviceCtrl) -> int:
+            logging.exception('Error running OpenCommissioningWindow %s', e)
+            asserts.fail('Failed to open commissioning window')
+
+    async def open_commissioning_window_with_subscription_monitoring(
```
We are now using this function in TC_CADMIN_1_3_4 test module test step 9 as this was intended to be used there previously, I just appear to have missed updating this previously, my apologies. Thank you for catching this!

j-ororke commented on src/python_testing/TC_CADMIN_1_3_4.py:
```diff
@@ -182,59 +182,50 @@ async def combined_commission_val_steps(self, commission_type: str):
             self.step(9)
             # TH_CR2 opens a commissioning window on DUT_CE for 180 seconds using ECM
             await self.th2.OpenCommissioningWindow(nodeid=self.dut_node_id, timeout=180, iteration=1000, discriminator=0, option=1)
+            results = await self.monitor_commissioning_window_closure_with_subscription(
```
Hi Cecille,
Thank you for catching this race condition! 
We have now updated the code to use open_commissioning_window_with_subscription_monitoring which starts the subscription first, then opens the commissioning window - eliminating the race condition you identified. We then use the existing subscription with monitor_commissioning_window_closure_with_subscription to verify the window closes within the expected timeframe. This ensures we capture all status changes from the moment the window opens.

j-ororke committed:
Updating TC_CADMIN_1_3_4 test module and cadmin_support modules:
- Fixed the race condition in step 9 by using open_commissioning_window_with_subscription_monitoring to establish the subscription before opening the commissioning window, then using the monitor_commissioning_window_closure_with_subscription function to monitor the subscription until the commissioning window closes.
- Fixed the AttributeError by replacing ._jsonData["sdk-config"]["f/1/r"] with the proper API call GetSdkKey("f/1/r")
- Function wait_for_window_status_change now asserts on timeout, otherwise silently passes without a return as not needed since we know it passes if it doesnt experience a timeout
- Removed the dead code from the log_timing_results function after changes made to wait_for_window_status_change function
- Removed open_commissioning_window_with_full_args function as it is best to keep the code simple adding without unneccessary functionality.

j-ororke committed:
Merge branch 'master' into create_CADMIN_clock-skew

restyled-commits committed:
Restyled by isort

j-ororke committed:
Resolving linting issues and restyled by isort

j-ororke committed:
Merge branch 'master' into create_CADMIN_clock-skew

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
Merge branch 'master' into create_CADMIN_clock-skew

restyled-commits committed:
Restyled by autopep8

j-ororke commented on src/python_testing/support_modules/cadmin_support.py:
```diff
@@ -134,7 +403,21 @@ def matches(self, expected_cm: int, expected_d: int) -> bool:
             return cm_match and d_match
 
-    async def wait_for_correct_cm_value(self, expected_cm_value: int, expected_discriminator: int, max_attempts: int = 5, delay_sec: int = 5):
+    async def get_all_txt_records(self):
```
Thank you for catching this, 
This function is not currently used in the CADMIN tests, so it must have been a merge issue and removed from the other tests when Raul was working on his mDNS changes PR, he had went through the changes he made for the CADMIN tests when he did those.
I have now removed this function and it should no longer be included.
Thanks again for catching this!

j-ororke committed:
Updating cadmin_support and TC_CADMIN_1_3_4 python3 test modules:
- Removed unused get_all_txt_records function
- Updated to using TimingResults dataclass instead of dictionary

j-ororke committed:
Resolving linting error

j-ororke commented on src/python_testing/support_modules/cadmin_support.py:
```diff
@@ -58,6 +73,260 @@ async def read_currentfabricindex(self, th: ChipDeviceCtrl) -> int:
+            )
+
+            # Log results
+            self.log_timing_results(results)
```
Sorry, you are correct, appears I did forget to remove the previous error logging on lines 250-255, I have removed that block of code here now. Thank you for catching that!

j-ororke committed:
Removed duplicated error logging for timing results in cadmin_support module

j-ororke committed:
Merge branch 'master' into create_CADMIN_clock-skew
